nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+    pub fn try_send(&mut self, packet: &Packet) -> Result<(), ()> {
+        // NOTE we do NOT check the address of `packet`; see comment in `Packet::new` for details
+        // go to the RXIDLE state
+        self.enable_rx();
+
+        // start CCA
+        self.radio
+            .tasks_ccastart
+            .write(|w| w.tasks_ccastart().set_bit());
+
+        loop {
+            if self
+                .radio
+                .events_ccaidle
+                .read()
+                .events_ccaidle()
+                .bit_is_set()
+            {
+                // channel is clear
+                self.radio.events_ccaidle.reset();
+                break;
+            }
+
+            if self
+                .radio
+                .events_ccabusy
+                .read()
+                .events_ccabusy()
+                .bit_is_set()
+            {
+                // channel is busy
+                self.radio.events_ccaidle.reset();
+                return Err(());
+            }
+        }
+
+        // NOTE(unsafe) DMA transfer has not yet started
+        unsafe {
+            self.radio
+                .packetptr
+                .write(|w| w.packetptr().bits(packet.buffer.as_ptr() as u32));
+        }
+
+        // the DMA transfer will start at some point after the following write operation so we place
+        // the barrier here
+        dma_start_fence();
+        // enter TX mode
+        self.radio.tasks_txen.write(|w| w.tasks_txen().set_bit());
+
+        // due to a shortcut the transmission will start automatically so we just have to wait
+        // until the PHYEND event is raised
+        self.wait_for_event(Event::PhyEnd);
+        dma_end_fence();
+
+        Ok(())
+    }
+
+    /// Sends the given `packet`
+    ///
+    /// This is utility method that *consecutively* calls the `try_send` method until it succeeds.
+    /// Note that this approach is *not* IEEE spec compliant -- there must be delay between failed
+    /// CCA attempts to be spec compliant
+    pub fn send(&mut self, packet: &Packet) {
+        // NOTE we do NOT check the address of `packet`; see comment in `Packet::new` for details
+
+        self.radio.shorts.modify(|_, w| w.ccaidle_txen().set_bit());
+
+        // go to the RXIDLE state
+        self.enable_rx();
+
+        // NOTE the DMA doesn't exactly start at this point but due to shortcuts it may occur at any
+        // point after this volatile write
+        dma_start_fence();
+        // NOTE(unsafe) DMA transfer has not yet started
+        unsafe {
+            self.radio
+                .packetptr
+                .write(|w| w.packetptr().bits(packet.buffer.as_ptr() as u32));
+        }
+
+        'cca: loop {
+            // start CCA
+            self.radio
+                .tasks_ccastart
+                .write(|w| w.tasks_ccastart().set_bit());
+
+            loop {
+                if self
+                    .radio
+                    .events_ccaidle
```

Comment:
You could wait for PHYEND instead. There is no need to wait here for CCAIDLE and for PHYEND later